Fix ref-counting when CompositeByteBuf is used with retainedSlice()

njhill opened the pull request:
Motivation:

`ByteBuf.retainedSlice()` and similar methods produce sliced buffers with an independent refcount to the buffer that they wrap.


One of the optimizations in 10539f4dc738c48e8d63c46b72ca32906d7f40ec was to use a ref to the unwrapped buffer object for added slices, but this did not take into account the above special case when later releasing.

Thanks to @rkapsi for discovering this via #8495.

Modifications:

Since a reference to the slice is still kept in the `Component` class, just changed `Component.freeIfNecessary()` to release the slice in preference to the unwrapped buf.

Also added a unit test which reproduces the bug.

Result:

Fixes #8495

normanmaurer commented:
@netty-bot test this please

rkapsi commented:
Took the branch for a spin, no more Exceptions or ERRORs as reported in #8495 .

normanmaurer commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -1815,7 +1815,11 @@ ByteBuf duplicate() {
+            // refcount to the unwrapped buf if it is a PooledSlicedByteBuf
+            ByteBuf s = slice;
+            (s != null ? s : buf).release();
+            // null out in both cases since it could be racy
```
I think this would be easier to read like this:

```
ByteBuf buffer = s == null ? buf : s;
buffer.release();
```

normanmaurer replied:
@njhill I am confused... why did you need to to have `s` and `toRelease` here ? Which not just do what I suggested ? 

Which would be:

```
ByteBuf buffer = slice == null ? buf : slice;
buffer.release();
```

njhill replied:
@normanmaurer oh, sorry... I misinterpreted your original suggestion since it used `s`. Maybe I am being overly cautious, but just based on the fact that `slice` gets nulled after the component is released I thought it would be safer to do the check on a local var (I ack that it probably is only possible to hit an NPE problem if there is another bug or incorrect use though).

Happy to change it to use `slice` directly if you think that's fine...

normanmaurer replied:
yeah I think I would go for the easier code or would at least not write it so compact to make it easier to understand

njhill replied:
done - changed to use if/else, hopefully this is the most readable of all :)

normanmaurer commented on buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java:
```diff
@@ -1144,6 +1144,36 @@ public void testReleasesItsComponents() {
 
+    @Test
+    public void testReleasesItsComponents2() {
+        ByteBuf buffer = PooledByteBufAllocator.DEFAULT.buffer(); // 1
```
Can you add a comment that using the PooledByteBufAllocator is important here ?

normanmaurer commented on buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java:
```diff
@@ -1144,6 +1144,36 @@ public void testReleasesItsComponents() {
+        ByteBuf s3 = s2.readRetainedSlice(2); // 4
+        ByteBuf s4 = s3.readRetainedSlice(2); // 5
+
+        ByteBuf composite = PooledByteBufAllocator.DEFAULT.compositeBuffer()
```
same as above

njhill replied:
@normanmaurer Actually I don't think this one would make a difference since it's only used if/when the composite buffer gets expanded right?

normanmaurer replied:
@njhill ah yeah true... then fix it please to use Unpooled.

normanmaurer commented on buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java:
```diff
@@ -1144,6 +1144,36 @@ public void testReleasesItsComponents() {
+        assertEquals(2, buffer.refCnt());
+
+        // releasing composite should release the 4 components
+        ReferenceCountUtil.release(composite);
```
just use `composite.release()`

normanmaurer commented on buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java:
```diff
@@ -1144,6 +1144,36 @@ public void testReleasesItsComponents() {
+        assertEquals(1, buffer.refCnt());
+
+        // last remaining ref to buffer
+        ReferenceCountUtil.release(buffer);
```
just use `buffer.release()`

normanmaurer commented:
@rkapsi thanks a lot!

njhill commented:
Thanks @normanmaurer, have pushed a commit addressing your comments.

Re the use of `ReferenceCountUtil.release(buf)` and pooled/unpooled `compositeBuffer()` - they seem to be used in many of the other tests, here I just copy/pasted the test above and changed `readSlice().retain()` to `readRetainedSlice()`. Just FYI that there's some inconsistency.

normanmaurer commented:
@njhill I guess it would make sense to cleanup this stuff as a followup.

normanmaurer commented:
@njhill also see https://github.com/netty/netty/pull/8497/files#r233174008

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@njhill merged... thanks!
